From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Cache world border

The world border is only initialized once when the level is created.
Lookup from data storage map multiple time is quite unnecessary and expensive.

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 8358864ab518c04af208bfb64e6362780c3a1b04..6c26eb69cbe7ae0e732e1c89a2fddcc94a3e5707 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -245,6 +245,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public boolean hasRidableMoveEvent = false; // Purpur - Ridables
     final List<ServerPlayer> realPlayers; // Leaves - skip
     public final org.dreeam.leaf.async.tracker.AsyncTracker leaf$asyncTracker = new org.dreeam.leaf.async.tracker.AsyncTracker(this); // Leaf - Multithreaded tracker
+    private @Nullable WorldBorder worldBorder; // Leaf - Cache world border
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
@@ -2599,8 +2600,14 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     @Override
     public WorldBorder getWorldBorder() {
+        // Leaf start - Cache world border
+        if (this.worldBorder != null) {
+            return this.worldBorder;
+        }
+        // Leaf end - Cache world border
         WorldBorder worldBorder = this.getDataStorage().computeIfAbsent(WorldBorder.TYPE);
         worldBorder.applyInitialSettings(this.levelData.getGameTime());
+        this.worldBorder = worldBorder; // Leaf - Cache world border
         return worldBorder;
     }
 
